File: backend/src/main/java/com/finora/imports/ownership/HolderNameMatcher.java
package com.finora.imports.ownership;

import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.Locale;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * docs/proposals/account-ownership-intelligence-proposal.md §3.1: compares a statement's extracted
 * holder name against a Finora profile name. A token-set comparison, not edit distance -- real
 * variance in how a name prints (initials, honorifics, word order, a second joint holder) looks
 * nothing like a typo, and edit distance would either miss it or need per-case tuning to catch it.
 *
 * <p>Deliberately does not split the extracted name into separate candidate holders before matching
 * -- see the design doc's §3.1 point 3 for why a naive split is wrong: "RAHUL AND PRIYA SHARMA" split
 * into ["RAHUL", "PRIYA SHARMA"] would separate the shared surname from the first name it belongs to,
 * and neither half alone would satisfy a profile of "Rahul Sharma". Instead, the whole extracted
 * string is tokenized as one set (with "AND"/"OR" dropped as connector words, and "&" already absent
 * from a letters-only split), and a match requires every profile-name token to be satisfied by SOME
 * token in that set -- which handles a joint holder correctly without needing to know where the
 * joiner is.
 */
public final class HolderNameMatcher {

    private HolderNameMatcher() {}

    private static final Set<String> CONNECTOR_WORDS = Set.of("AND", "OR");

    /**
     * @return true if every token in {@code profileName} is satisfied (exact match, or as an
     * initial) by some token extracted from {@code extractedHolderName}. False if either name is
     * blank -- there is nothing to compare, and "nothing to compare" is a caller-level concern
     * ({@code NO_HOLDER_FOUND}), not a match.
     */
    public static boolean isLikelyMatch(String extractedHolderName, String profileName) {
        Set<String> extractedTokens = tokensOf(extractedHolderName);
        Set<String> profileTokens = tokensOf(profileName);
        if (extractedTokens.isEmpty() || profileTokens.isEmpty()) return false;

        for (String profileToken : profileTokens) {
            boolean satisfied = extractedTokens.stream().anyMatch(t -> tokenMatches(profileToken, t));
            if (!satisfied) return false;
        }
        return true;
    }

    private static Set<String> tokensOf(String name) {
        if (name == null) return Set.of();
        return Arrays.stream(name.trim().toUpperCase(Locale.ROOT).split("[^A-Z]+"))
                .filter(t -> !t.isBlank() && !CONNECTOR_WORDS.contains(t))
                .collect(Collectors.toCollection(LinkedHashSet::new));
    }

    /** Equal, or one is a single-letter initial of the other. Not commutative-by-accident -- both
     *  directions are checked because either side (statement or profile) might be the abbreviated
     *  one ("R Sharma" on the statement vs. "Rahul Sharma" on the profile, or the reverse). */
    private static boolean tokenMatches(String a, String b) {
        if (a.equals(b)) return true;
        if (a.length() == 1) return b.startsWith(a);
        if (b.length() == 1) return a.startsWith(b);
        return false;
    }
}

File: frontend/src/lib/holderNameMatcher.ts
/**
 * docs/proposals/account-ownership-intelligence-proposal.md §3.1: does a statement's extracted
 * holder name plausibly belong to the logged-in user's profile? A client-side port of
 * `HolderNameMatcher` (backend/src/main/java/com/finora/imports/ownership/HolderNameMatcher.java)
 * -- same token-set algorithm, same reasoning, kept in sync deliberately rather than shared,
 * since this side only ever gates whether to show the review dialog before confirming; the
 * backend's own comparison (via `OwnershipMatchService`) is what actually gets persisted and is
 * the source of truth. A rare disagreement between the two (e.g. this import turns out to resolve
 * to an existing account, which only the backend knows at confirm time) means at worst one
 * unnecessary dialog on an otherwise-legitimate import -- never a missed warning becoming a
 * silent block, since neither side ever blocks anything.
 *
 * Deliberately does not split the extracted name into separate candidate holders before matching
 * -- a joint statement's shared surname must stay associated with the first name it belongs to.
 * See the Java version's own doc comment for the worked example.
 */

const CONNECTOR_WORDS = new Set(['AND', 'OR']);

export function isLikelyMatch(extractedHolderName: string | null | undefined, profileName: string | null | undefined): boolean {
  const extractedTokens = tokensOf(extractedHolderName);
  const profileTokens = tokensOf(profileName);
  if (extractedTokens.size === 0 || profileTokens.size === 0) return false;

  for (const profileToken of profileTokens) {
    const satisfied = [...extractedTokens].some((t) => tokenMatches(profileToken, t));
    if (!satisfied) return false;
  }
  return true;
}

function tokensOf(name: string | null | undefined): Set<string> {
  if (!name) return new Set();
  const tokens = name
    .trim()
    .toUpperCase()
    .split(/[^A-Z]+/)
    .filter((t) => t.length > 0 && !CONNECTOR_WORDS.has(t));
  return new Set(tokens);
}

/** Equal, or one is a single-letter initial of the other. Checked both directions -- either the
 *  statement or the profile might be the abbreviated one. */
function tokenMatches(a: string, b: string): boolean {
  if (a === b) return true;
  if (a.length === 1) return b.startsWith(a);
  if (b.length === 1) return a.startsWith(b);
  return false;
}
